File: python/p1.py
def main():
    fname = '../data/p1.txt'
    f = open(fname, 'r')

    nums = [int(l) for l in f.readlines()]
    num_set = set()

    for n in nums:
        if (2020 - n) in num_set:
            print('Part one: {}'.format(n * (2020 - n)))
            break
        num_set.add(n)

    pair_sums = {}
    for n1 in nums:
        for n2 in nums:
            pair_sums[n1 + n2] = (n1, n2)

    for n3 in nums:
        if (2020 - n3) in pair_sums:
            n1, n2 = pair_sums[2020 - n3]
            print("Part two: {}".format(n1 * n2 * n3))
            break

if __name__ == '__main__':
    main()

File: julia/p1.jl
function get_nums()
    fname = "../data/p1.txt";
    word_nums = readlines(fname);
    nums = map(x->parse(Int, x), word_nums);
    return nums;
end

function part_one(nums)
    nums = Set(nums);

    for n in nums
        if in(2020 - n, nums)
            println("Part one: ", n * (2020 - n));
            return;
        end

        push!(nums, n);
    end
end

function part_two(nums)
    two_sums = Set();

    for n1 in nums
        for n2 in nums
            for n3 in nums
                if (n1 + n2 + n3) == 2020
                    println("Part two: ", n1 * n2 * n3);
                    return;
                end
            end
        end
    end
end

function main()
    nums = get_nums();
    part_one(nums);
    part_two(nums);
end

main();
